Reference content

oguzeroglu opened the pull request:
This PR proposes a reference content for:

* Module (BPMN modeling module)
* API reference (importXML)
* Command (element.updateLabel)
* Event (import.parse.start)
* Example (bpmn properties and theming)

oguzeroglu committed:
feat(API-Reference): importXML documentation

oguzeroglu committed:
feat(Events): document import-parse-start event

oguzeroglu committed:
feat(Examples): document bpmn-properties example

oguzeroglu committed:
feat(Examples): add Theming example

oguzeroglu committed:
feat(Modules): BPMN modeling module

barmac commented on docs/API-Reference/bpmn-js/importXML.md:
````diff
@@ -10,3 +10,62 @@ grand_parent: "API Reference"
+### Code Example
+
+```javascript
+const modeler = new BpmnJS();
````
The import statement is missing in the example.

pinussilvestrus replied:
You don't need an import statement every time, e.g. when you integrate the production builds directly inside the browser.

oguzeroglu replied:
I removed the BpmnJS construction code and added a comment that says "modeler is an instance of BpmnJS" instead since the BpmnJS constructor also expects the container DOM element. So it does not make sense to include those in every API Reference code snippets. Let's make the code snippets focus on only the API.

barmac commented on docs/API-Reference/bpmn-js/importXML.md:
```diff
@@ -10,3 +10,62 @@ grand_parent: "API Reference"
+
+try {
+
+  const result = await modeler.importXML(xml);
```
I'd love the example to be runnable. With top-level `await`, it's not yet available.

MaxTru replied:
What about switching to `.then(...)`?

philippfromme replied:
The question is if users should be able to copy and paste example code that works out of the box. If that's the case we can't do stuff like `new BpmnJS()` without specifying a container since that will error.

oguzeroglu replied:
Yeah I don't think the code could be runnable without specifiying the DOM element and import statements. It does not make sense to include those in every code snippet IMO. I also removed the modeler construction code and added a comment instead. In this context, await is fine here IMO. It makes it clear that this API is awaitable / returns a Promise.

philippfromme replied:
We cannot include all the setup code in every example so we shouldn't try. We should rather decide on the naming that we use throughout the docs. Are we going to use `viewer` and `modeler`? Or `bpmnJS`?

MaxTru replied:
:+1:

oguzeroglu replied:
I updated the comment with: `modeler is an instance of BpmnJS Modeler.` So at that point it's clear that we're using a Modeler for given example. Wheter one should use a modeler or a viewer should become clear via other pages such as `Getting Started`. It's not the concern of the importXML API reference page.

barmac commented on docs/Events/bpmn-events/import-parse-start.md:
````diff
@@ -10,3 +10,25 @@ grand_parent: "Events"
+## Example Usage
+
+```javascript
+bpmnJS.on('import.parse.start', function(data) {
+
+  const { xml } = data; // this is the XML that is about to get imported
+
+  const newXML = replace('StartEvent_1', 'StartEvent_2');
+
+  return newXML; // now this new XML will be imported instead
+});
````
If we run this example, the import will be broken due to missing BPMN element for StartEvent_1 DI.

oguzeroglu replied:
This is just an example to demonstrate what this event does. It's hard to tell what is broken since we don't know whats inside the XML anyway. We can instead add a regex replace all kinda thing there but it's not adding more value to it IMO. What would be the alternative that you propose instead of this code?

barmac replied:
It could be sth like `replace('Order pizza', 'Order sushi')` (clearly a label and not ID), but I'm also OK with how it looks now.

barmac commented on docs/Modules/bpmn-modules/modeling.md:
````diff
@@ -9,4 +9,160 @@ parent: "BPMN Modules"
+### Usage
+
+```javascript
+const elementRegistry = bpmnJS.get('elementRegistry');
+const modeling = bpmnJS.get('modeling');
+
+const startEvent_1 = elementRegistry.get('StartEvent_1');
+
+modeling.updateLabel(startEvent_1, 'new label');
````
Cool example. I like it!

pinussilvestrus commented on docs/API-Reference/bpmn-js/importXML.md:
```diff
@@ -10,3 +10,62 @@ grand_parent: "API Reference"
+
+With the [v7.0.0 release](https://github.com/bpmn-io/bpmn-js/blob/develop/CHANGELOG.md#700) public bpmn-js APIs are made awaitable and passing callbacks to APIs will be deprecated in a future release. See [here](https://bpmn.io/l/moving-to-promises.html) for more info.
+
+# ImportXMLResult
```
Are we introducing these types with the docs or are they defined inside our codebase?

philippfromme replied:
Not sure whether we should invent type names here. People might think that these are TypeScript types.

oguzeroglu replied:
I gave this topic some thought. We do the same in the documentation in our codebase to define certain types. It's important to state what the Promise resolves exactly so we should include types IMO. We don't claim we support Typescript anywhere and Typescript is yet-another-3rd-party-Javascript-related-tool which is pretty popular but it's still a 3rd party tool. I don't think we should go that far to shape the documented content based on a 3rd party tool.

One additional thing I like about this approach is that the users can click on the types and go to the type documentation via permalinks. So it's pretty useful.

What would be the alternative you propose if you want to chance our approach here?

philippfromme replied:
We use types in the code for things that like `djs.model.Base`, `ModdleElement` and `Event`. These things are actual prototypes, not one-off things. So I'd like us to find another solution for documenting this.

oguzeroglu replied:
Took care of this one 👍

pinussilvestrus commented on docs/Modules/bpmn-modules/modeling.md:
```diff
@@ -9,4 +9,160 @@ parent: "BPMN Modules"
+* [splitLane](#splitlane)
+* [setColor](#setcolor)
+
+## updateLabel
```
I miss a bit the general input parameter and return value information to use this API. Or do we describe them in the separate command sections?

philippfromme replied:
Parameters of each method should be documented.

oguzeroglu replied:
Done 👍

pinussilvestrus commented on docs/Modules/bpmn-modules/modeling.md:
````diff
@@ -9,4 +9,160 @@ parent: "BPMN Modules"
+
+To create connection with custom start/end points:
+```javascript
+modeling.connect(sourceShape, targetShape, null, {
````
As a user, I now wonder what the "null" parameter is about. How to know what parameters to fill inside the API method?

philippfromme replied:
Agreed. We shouldn't use `null` without any explanation.

Could be changed to

```javascript
const attrs = null;

modeling.connect(sourceShape, targetShape, attrs, { ... });
```

oguzeroglu replied:
Updated the example, also now the input/output params of each API is included so that'd not be a problem anymore.

pinussilvestrus commented on docs/Modules/bpmn-modules/modeling.md:
```diff
@@ -9,4 +9,160 @@ parent: "BPMN Modules"
 
-# modeling
\ No newline at end of file
+# modeling
```
I very like the examples 👏

oguzeroglu replied:
Thanks 🎸

pinussilvestrus reviewed (commented):
Did we already decide what to do with empty pages?

MaxTru commented:
> Did we already decide what to do with empty pages?

Yes, we will hide them on the docs page by removing the top-level config section in the `.md` file.

MaxTru commented on docs/API-Reference/bpmn-js/importXML.md:
```diff
@@ -10,3 +10,62 @@ grand_parent: "API Reference"
+
+## Usage
+
+importXML(xml: _[String](#)_, bpmnDiagram: _[ModdleElement](#)_ or _[String](#)_): Promise<_[ImportXMLResult](#importxmlresult)_, _[ImportXMLError](#importxmlerror)_>
```
_[String](#)_ doesn't require a link IMO, so it can be just _`String`_

oguzeroglu replied:
Ah yeah good idea. Updated the empty links in input parameters.

oguzeroglu replied:
Same goes for the ModdleElement parameter since it does not point to any page.

MaxTru commented on docs/API-Reference/bpmn-js/importXML.md:
```diff
@@ -10,3 +10,62 @@ grand_parent: "API Reference"
+
+importXML(xml: _[String](#)_, bpmnDiagram: _[ModdleElement](#)_ or _[String](#)_): Promise<_[ImportXMLResult](#importxmlresult)_, _[ImportXMLError](#importxmlerror)_>
+
+* `xml`: The BPMN 2.0 XML.
```
Style of parameter description I suggest to do formulate in format "<what> <type> (optionally)to <do what>"

So here:
BPMN 2.0 XML string to import and render.

oguzeroglu replied:
Done 👍

MaxTru commented on docs/API-Reference/bpmn-js/importXML.md:
```diff
@@ -10,3 +10,62 @@ grand_parent: "API Reference"
+importXML(xml: _[String](#)_, bpmnDiagram: _[ModdleElement](#)_ or _[String](#)_): Promise<_[ImportXMLResult](#importxmlresult)_, _[ImportXMLError](#importxmlerror)_>
+
+* `xml`: The BPMN 2.0 XML.
+* `bpmnDiagram`: The BPMN diagram or the id of the diagram to render _(if not provided, the first one will be rendered)_.
```
See above.

BPMN diagram moddle element to render **or** id string of the diagram to render.

(I am not sure I understand this second parameter)

oguzeroglu replied:
Updated the explanation. It seems like some BPMN XML could contain multuple diagrams, so in this case this 2nd parameter helps us to specify which diagram to import.

MaxTru commented on docs/API-Reference/bpmn-js/importXML.md:
```diff
@@ -10,3 +10,62 @@ grand_parent: "API Reference"
+During the import, these [events](/events) will be fired:
+
+* [import.parse.start](/events/bpmn-events/import-parse-start): About to read model from xml
+* [import.parse.complete](/events/bpmn-events/import-parse-complete): Model is read, either successfully or with errors.
```
Model is parsed[...]

MaxTru commented on docs/API-Reference/bpmn-js/importXML.md:
```diff
@@ -10,3 +10,62 @@ grand_parent: "API Reference"
+
+During the import, these [events](/events) will be fired:
+
+* [import.parse.start](/events/bpmn-events/import-parse-start): About to read model from xml
```
Let's discuss: I believe the event names are speaking for themselves, the descriptions at the end of each line in the next esection can therefore be removed [since currently they bascailly just say the same thing as the name, but with different terms]

philippfromme replied:
Since the event names link to the details of the events we can remove the explanation here.

oguzeroglu replied:
Yeah it makes sense. I removed the explanations 👍

MaxTru commented on docs/API-Reference/bpmn-js/importXML.md:
```diff
@@ -10,3 +10,62 @@ grand_parent: "API Reference"
+
+With the [v7.0.0 release](https://github.com/bpmn-io/bpmn-js/blob/develop/CHANGELOG.md#700) public bpmn-js APIs are made awaitable and passing callbacks to APIs will be deprecated in a future release. See [here](https://bpmn.io/l/moving-to-promises.html) for more info.
+
+# ImportXMLResult
```
Shouldn't this be in another `.md`? OR in a sub-section of `# importXML` named `## Returns` or something like that?

oguzeroglu replied:
I added this here for these 2 reasons:

1) We want permalink functionality: When you click on the ImportXMLResult link in the input/output parameters section, the page would scroll down to this header.
2) This is not a shared type, we only use this pseudo-type for the importXML API. We use this same pseudo-type in our codebase while documenting the importXML API.

oguzeroglu replied:
Does this make sense?

MaxTru replied:
Yeah, makes sense. Ty

MaxTru commented on docs/Examples/bpmn-examples/bpmn-properties.md:
```diff
@@ -10,3 +10,7 @@ grand_parent: "Examples"
 ---
```
Let's discuss how to deal with examples.

We will have multiple entry points to our examples:
https://bpmn.io/toolkit/bpmn-js/examples/
https://github.com/bpmn-io/bpmn-js-examples
And then we would another entry point: the docs.

oguzeroglu replied:
So I suggest keeping the examples here as we want them findable/searcheable. A user who searches for the term "theme" could find the theming example in the docs webpage. This is not the case for the examples Github repo.

One additional change I'll do is: As Philipp stated, it makes sense to include the README of the example content to the example docs pages since it'll help it to be more visible in the search results.

philippfromme replied:
I think it's okay to have all of these. GitHub is where the actual code of the examples resides. On our bpmn.io website, we use previews of these examples to give the user an idea of what might be possible with the toolkit. These previews merely link to GitHub, they could instead link to the docs. And then the docs contain the documentation part of the examples (in most cases what is now the README on GitHub) and link to the actual code on GitHub.

MaxTru commented on docs/Examples/bpmn-examples/theming.md:
```diff
@@ -10,3 +10,9 @@ grand_parent: "Examples"
 ---
 
 # Theming
```
Should we not rather have one page which lists all examples? Since it would be easier then for the user to get an overview. Maybe only this: https://github.com/bpmn-io/bpmn-js-examples/README.md embedded

Let's dicuss

oguzeroglu replied:
We actually have it here: https://bpmn-io.github.io/docs.bpmn.io/examples/bpmn

oguzeroglu replied:
We also have the overview page for every section. This is a default feature by the Jekyll plugin that we're using.

philippfromme commented on docs/API-Reference/bpmn-js/importXML.md:
```diff
@@ -10,3 +10,62 @@ grand_parent: "API Reference"
+
+The importXML error.
+
+## Properties
```
`message` property is missing.

oguzeroglu replied:
Done 👍

philippfromme commented on docs/Commands/bpmn-commands/element-updateLabel.md:
```diff
@@ -10,3 +10,5 @@ grand_parent: "Commands"
 ---
 
 # element.updateLabel
+
```
I'd say, either document the command or leave it for now.

oguzeroglu replied:
Ehm, I don't quite get this comment. It's already documented?

philippfromme replied:
Saying that this is the command that is executed by some method is not really that helpful. It could be executed by anyone trough the command stack. What would be helpful would be to know what context will be available at what phase of the command execution. But I'm not sure whether we should invest in documenting all these details now.

oguzeroglu replied:
Ok updated the documentation and added the context information. For this page for instance, added the `oldLabel` as `Post execution only`

philippfromme commented on docs/Events/bpmn-events/import-parse-start.md:
```diff
@@ -10,3 +10,25 @@ grand_parent: "Events"
+
+  const { xml } = data; // this is the XML that is about to get imported
+
+  const newXML = replace('StartEvent_1', 'StartEvent_2');
```
We shouldn't include imaginary functions in the examples. I don't know what exactly `replace` is doing.

oguzeroglu replied:
Yeah it was supposed to be `xml.replace`, I fixed that 👍

philippfromme commented on docs/Events/bpmn-events/import-parse-start.md:
````diff
@@ -10,3 +10,25 @@ grand_parent: "Events"
+## Example Usage
+
+```javascript
+bpmnJS.on('import.parse.start', function(data) {
````
If we use `const` we should use arrow functions, too.

oguzeroglu replied:
Done 👍

philippfromme commented on docs/Examples/bpmn-examples/bpmn-properties.md:
```diff
@@ -10,3 +10,7 @@ grand_parent: "Examples"
 ---
 
 # BPMN Properties
```
The example should contain the full example (README of the repository) _and_ additionally link to the code. Otherwise having a full page per example is a waste.

oguzeroglu replied:
Done 👍 I updated the Theming and Bpmn Properties examples. It makes sense since more text we have -> more visible in search

MaxTru replied:
So we duplicate the `README.md` of the example. I don't really like that. 
Is there not maybe a jekyll feature to embed / just update to a given uri (which would be the `README.md`)?

oguzeroglu replied:
This is a good point Max. I found a Jekyll plugin for this but we still need to spend some time to investigate wheter it is possible to use this plugin with the Github pages + Just-The-Docs theme. I opened this ticket about this topic: https://github.com/bpmn-io/docs.bpmn.io/issues/27

philippfromme commented on docs/Modules/bpmn-modules/modeling.md:
```diff
@@ -9,4 +9,160 @@ parent: "BPMN Modules"
\ No newline at end of file
+# modeling
+
+This module activates the BPMN 2.0 modeling features [over the common modeling module](/modules/common-modules/modeling). It means that this module contains every API exported by the common modeling module.
```
Not sure what _activates features over the common modeling module_ means. Let's say the module extends the functionality of the existing diagram-js modeling module.

oguzeroglu replied:
Done 👍

philippfromme commented on docs/Modules/bpmn-modules/modeling.md:
```diff
@@ -9,4 +9,160 @@ parent: "BPMN Modules"
+const elementRegistry = bpmnJS.get('elementRegistry');
+const modeling = bpmnJS.get('modeling');
+
+const startEvent_1 = elementRegistry.get('StartEvent_1');
```
We should stick with our code style and use camelCase in favor of snake_case.

oguzeroglu replied:
Done 👍

philippfromme commented on docs/Modules/bpmn-modules/modeling.md:
```diff
@@ -9,4 +9,160 @@ parent: "BPMN Modules"
+
+## connect
+
+`connect` API performs the same job as [the common createConnection API](/modules/common-modules/modeling#createconnection). The only difference is that `connect` API first checks if the source element and the target element are connectable by BPMN standards before delegating to `createConnection` API.
```
Starting every explanation with the name of the method is redundant.

philippfromme replied:
Wording should be improved.

Maybe

```
Connects two shapes if allowed by the [BPMN rules](/modules/bpmn-modules/bpmn-rules). Uses [createConnection](/modules/common-modules/modeling#createconnection) internally.
```

oguzeroglu replied:
Done 👍

philippfromme reviewed (changes requested):
Maybe we should reconsider the structure of the docs. bpmn-js and dmn-js are quite mixed up. I think in most cases people want to focus on one toolkit at a time. In the case of bpmn-js they should easily see the diagram-js and bpmn-js docs. So the top-level structure could by divided by library. diagram-js docs would then also show up in bpmn-js and dmn-js docs. We have to be smart about this.

oguzeroglu commented:
@philippfromme I don't quite get this. Could you write an example top level structure? What's on your mind?

oguzeroglu commented:
Philipp and I decided to deal with the top-level structure later. It does not block us from merging the reference content. So we can do another round of review on this. I addressed your comments.

MaxTru commented:
> Philipp and I decided to deal with the top-level structure later. It does not block us from merging the reference content. So we can do another round of review on this. I addressed your comments.

Added some minor comments - apart from them this looks cool to me!

philippfromme commented:
Regarding the structure of the documentation my idea is something along these lines:

## Example 1: bpmn-js selected

* Introduction
	* modules
	* commands
	* events
* Reference
	* BpmnJS (e.g. #clear, #destroy, #importXML, #saveXML)
		* EventBus (e.g. #on, #off)
		* Canvas (e.g. #getRootElement, #zoom)
		* Modeling (e.g. #createShape, #updateProperties)
		* BpmnReplace (e.g. #replaceElement)
* Examples

## Example 2: diagram-js selected

* Introduction
	* modules
	* commands
	* events
* Reference
	* Diagram (e.g. #clear, #destroy)
		* EventBus (e.g. #on, #off)
		* Canvas (e.g. #getRootElement, #zoom)
		* Modeling (e.g. #createShape)
		* Replace (e.g. #replaceElement)
* Examples

You'd avoid confusion by selecting the toolkit first and then seeing only the documentation that's relevant for the chosen toolkit. Choose diagram-js, see all diagram-js-related docs, choose bpmn-js, see all diagram-js-related _and_ bpmn-js-related docs.
